Mathlib/AlgebraicGeometry/RationalMap.lean

```diff
@@ -0,0 +1,308 @@
+  exact ⟨e, H⟩
+
+/-- The restriction of a partial map to a smaller domain. -/
+@[simps hom, simps (config := .lemmasOnly) domain]
+noncomputable
+def restrict (f : X.PartialMap Y) (U : X.Opens)
+    (hU : Dense (U : Set X)) (hU' : U ≤ f.domain) : X.PartialMap Y where
+  domain := U
+  dense_domain := hU
+  hom := X.homOfLE hU' ≫ f.hom
+
+@[simp]
+lemma restrict_id (f : X.PartialMap Y) : f.restrict f.domain f.dense_domain le_rfl = f := by
+  ext1
+  · rfl
+  simp
+
+/-- The composition of a partial map and a morphism on the right. -/
+@[simps]
+def compHom (f : X.PartialMap Y) (g : Y ⟶ Z) : X.PartialMap Z where
+  domain := f.domain
+  dense_domain := f.dense_domain
+  hom := f.hom ≫ g
+
+/-- A scheme morphism as a partial map. -/
+@[simps]
+def _root_.AlgebraicGeometry.Scheme.Hom.toPartialMap (f : X.Hom Y) :
+    X.PartialMap Y := ⟨⊤, dense_univ, X.topIso.hom ≫ f⟩
+
+/-- If `x` is in the domain of a partial map `f`, then `f` restricts to a map from `Spec 𝒪_x`. -/
+noncomputable
+def fromSpecStalkOfMem (f : X.PartialMap Y) {x} (hx : x ∈ f.domain) :
+    Spec (X.presheaf.stalk x) ⟶ Y :=
+  f.domain.fromSpecStalkOfMem x hx ≫ f.hom
+
+/-- A partial map restricts to a map from `Spec K(X)`. -/
+noncomputable
+abbrev fromFunctionField [IrreducibleSpace X] (f : X.PartialMap Y) :
+    Spec X.functionField ⟶ Y :=
+  f.fromSpecStalkOfMem
+    ((genericPoint_specializes _).mem_open f.domain.2 f.dense_domain.nonempty.choose_spec)
+
+lemma fromSpecStalkOfMem_restrict (f : X.PartialMap Y)
+    {U : X.Opens} (hU : Dense (U : Set X)) (hU' : U ≤ f.domain) {x} (hx : x ∈ U) :
+    (f.restrict U hU hU').fromSpecStalkOfMem hx = f.fromSpecStalkOfMem (hU' hx) := by
+  dsimp only [fromSpecStalkOfMem, restrict, Scheme.Opens.fromSpecStalkOfMem]
+  have e : ⟨x, hU' hx⟩ = (X.homOfLE hU').base ⟨x, hx⟩ := by
+    rw [Scheme.homOfLE_base]
+    rfl
+  rw [Category.assoc, ← Spec_map_stalkMap_fromSpecStalk_assoc,
+    ← Spec_map_stalkSpecializes_fromSpecStalk (Inseparable.of_eq e).specializes,
+    ← TopCat.Presheaf.stalkCongr_inv _ (Inseparable.of_eq e)]
+  simp only [← Category.assoc, ← Spec.map_comp]
+  congr 3
+  rw [Iso.eq_inv_comp, ← Category.assoc, IsIso.comp_inv_eq, IsIso.eq_inv_comp,
+    stalkMap_congr_hom _ _ (X.homOfLE_ι hU').symm]
+  simp only [restrictFunctor_obj_left, homOfLE_leOfHom, TopCat.Presheaf.stalkCongr_hom]
+  rw [← stalkSpecializes_stalkMap_assoc, stalkMap_comp]
+
+lemma fromFunctionField_restrict (f : X.PartialMap Y) [IrreducibleSpace X]
+    {U : X.Opens} (hU : Dense (U : Set X)) (hU' : U ≤ f.domain) :
+    (f.restrict U hU hU').fromFunctionField = f.fromFunctionField :=
+  fromSpecStalkOfMem_restrict f _ _ _
+
+/--
+Given `S`-schemes `X` and `Y` such that `Y` is locally of finite type and
+`X` is irreducible germ-injective at `x` (e.g. when `X` is integral),
+any `S`-morphism `Spec 𝒪ₓ ⟶ Y` spreads out to a partial map from `X` to `Y`.
+-/
+noncomputable
+def ofFromSpecStalk [IrreducibleSpace X] [LocallyOfFiniteType sY] {x : X} [X.IsGermInjectiveAt x]
```

Comment:
```suggestion
noncomputable
def ofFromSpecStalk [PreirreducibleSpace X] [LocallyOfFiniteType sY] {x : X} [X.IsGermInjectiveAt x]
```
and below.

Reply:
I don't think this matters because we assumed `X` is non-empty anyways (`by the `x : X`).

Reply:
Oh yes, I did not pay attention.